linera-chain/src/data_types.rs

```diff
@@ -421,17 +422,56 @@ pub struct LiteValue {
 struct ValueHashAndRound(CryptoHash, Round);
 
 /// A vote on a statement from a validator.
-#[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct Vote {
-    pub value: HashedCertificateValue,
+#[derive(Clone, Debug, Serialize)]
+pub struct Vote<T> {
+    pub value: Hashed<T>,
     pub round: Round,
     pub validator: ValidatorName,
     pub signature: Signature,
 }
 
-impl Vote {
+impl<'de, T: DeserializeOwned + BcsHashable> Deserialize<'de> for Vote<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Vote<T>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Debug, Deserialize)]
+        struct VoteHelper<T> {
+            value: T,
+            round: Round,
+            validator: ValidatorName,
+            signature: Signature,
+        }
+
+        let VoteHelper {
+            value,
+            round,
+            validator,
+            signature,
+        } = VoteHelper::deserialize(deserializer)?;
+
+        Ok(Vote {
+            value: Hashed::new(value),
```

Comment:
Would it be equivalent to use `Hashed<T>` in `VoteHelper` directly, or even derive `Deserialize`, since `Hashed`'s `Deserialize` impl already computes the hash?

Reply:
Unfortunately, it won't work - `Serialize for Hashed<T>` has the additional trait bound requiring that `T: BscHashable` and the `T` on the `*Helper` is a different `T` from the outer one (I cannot reuse - compiler complains). If I add that bound on the inner struct, like this:
```rust
#[derive(Debug, Deserialize)]
struct VoteHelper<T: BcsHashable> {
```
then compiler complains about something else:
```
multiple `impl`s or `where` clauses satisfying `T: Deserialize<'_>` found
required for `<Vote<T> as Deserialize<'de>>::deserialize::_::<impl Deserialize<'de> for VoteHelper<T>>::deserialize::__Visitor<'de, T>` to implement `Visitor<'de>`
```

Reply:
I see. :dizzy_face: